extensions-core/druid-lookups/src/main/java/io/druid/server/lookup/PollingLookup.java

```diff
@@ -0,0 +1,243 @@
+
+import javax.validation.constraints.NotNull;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class PollingLookup extends LookupExtractor
+{
+  private static final Logger LOGGER = new Logger(PollingLookup.class);
+
+  private final Period pollPeriod;
+
+  private final DataFetcher dataFetcher;
+
+  private final PollingCacheFactory cacheFactory;
+
+  private final AtomicReference<CacheRefKeeper> refOfCacheKeeper = new AtomicReference<>();
+
+
+  private final ListeningScheduledExecutorService scheduledExecutorService;
+
+  private final AtomicBoolean isOpen = new AtomicBoolean(false);
+
+  private final ListenableFuture<?> pollFuture;
+
+  public PollingLookup(
+      Period pollPeriod,
+      DataFetcher dataFetcher,
+      PollingCacheFactory cacheFactory
+  )
+  {
+
+    this.pollPeriod = pollPeriod == null ? Period.ZERO : pollPeriod;
+    this.dataFetcher = Preconditions.checkNotNull(dataFetcher);
+    this.cacheFactory = cacheFactory == null ? new OnHeapPollingCache.OnHeapPollingCacheProvider() : cacheFactory;
+    refOfCacheKeeper.set(new CacheRefKeeper(this.cacheFactory.makeOf(dataFetcher.fetchAll())));
+    if (getPollMs() > 0) {
+      scheduledExecutorService = MoreExecutors.listeningDecorator(Executors.newSingleThreadScheduledExecutor(
+          new ThreadFactoryBuilder()
+              .setDaemon(true)
+              .setNameFormat("PollingLookup-%d")
+              .setPriority(Thread.MIN_PRIORITY)
+              .build()
+      ));
+      pollFuture = scheduledExecutorService.scheduleAtFixedRate(
+          pollAndSwap(),
+          getPollMs(),
+          getPollMs(),
+          TimeUnit.MILLISECONDS
+      );
+    } else {
+      scheduledExecutorService = null;
+      pollFuture = null;
+    }
+    this.isOpen.set(true);
+  }
+
+
+  public void close()
+  {
+    LOGGER.info("Closing lookup");
+    synchronized (isOpen) {
+      isOpen.getAndSet(false);
+      CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.getAndSet(null);
+      if (cacheRefKeeper != null) {
+        cacheRefKeeper.doneWithIt();
+      }
+      if (pollFuture != null) {
+        pollFuture.cancel(true);
+        scheduledExecutorService.shutdown();
+      }
+    }
+  }
+
+  @Override
+  public String apply(@NotNull String key)
+  {
+    CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.get();
+    if (cacheRefKeeper == null) {
+      throw new ISE("Cache reference is null WTF");
+    }
+    PollingCache cache = cacheRefKeeper.get();
+    try {
+      if (cache == null) {
+        // it must've been closed after swapping while I was getting it.  Try again.
+        return this.apply(key);
+      }
+      return Strings.emptyToNull((String) cache.get(key));
+    }
+    finally {
+      if (cacheRefKeeper != null && cache != null) {
+        cacheRefKeeper.doneWithIt();
```

Comment:
If my recollection of the dimension extraction workflow is correct, the apply for the lookup extractor is called for multiple dimension values. As such this would get called N times for N dimension values getting looked-up. Is that intended here?